### PR TITLE
[#147] Optimize duplicate aggregation nodes

### PR DESCRIPTION
**Pull request by Seaven**

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleType.java [hunk 1]
```diff
@@ -77,6 +77,9 @@
 
     TF_MATERIALIZED_VIEW,
 
+    TF_MERGE_TWO_AGG_RULE,
+    TF_MERGE_TWO_AGG_WITH_PROJECT_RULE,
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleType.java:81]: Is this necessary?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        LogicalProjectOperator project = (LogicalProjectOperator) input.getInputs().get(0).getOp();
+        LogicalAggregationOperator aggregate2 =
+                (LogicalAggregationOperator) input.getInputs().get(0).getInputs().get(0).getOp();
+
+        // Only handle column prune, forbidden expression
+        if (!project.getColumnRefMap().entrySet().stream().allMatch(e -> e.getKey().equals(e.getValue()))) {
+            return false;
+        }
+
+        if (aggregate2.getPredicate() != null && aggregate2.hasLimit()) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:44]: There is OR ？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+import java.util.Map;
+
+public class MergeTwoAggRule extends TransformationRule {
+    private static final List<String> ALLOW_MERGE_AGGREGATE_FUNCTIONS = Lists.newArrayList(FunctionSet.SUM,
+            FunctionSet.MAX, FunctionSet.MIN, FunctionSet.MAX, FunctionSet.MIN, FunctionSet.BITMAP_UNION,
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:23]: redundant FunctionSet.MIN and FunctionSet.MAX

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+import java.util.Map;
+
+public class MergeTwoAggRule extends TransformationRule {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:21]: Need more comment for this class.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            return false;
+        }
+
+        for (CallOperator call : aggCallMap1.values()) {
+            ColumnRefOperator ref = (ColumnRefOperator) call.getChild(0);
+            if (!ALLOW_MERGE_AGGREGATE_FUNCTIONS.contains(call.getFnName().toLowerCase())) {
+                return false;
+            }
+
+            if (!aggCallMap2.containsKey(ref) && !FunctionSet.MAX.equalsIgnoreCase(call.getFnName()) &&
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:82]: Need comment  here

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggregate1 = (LogicalAggregationOperator) input.getOp();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:97]: `aggregateAbove` and `aggregateBelow`  is  better than `aggregate1` and `aggregate2`

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                    !FunctionSet.MIN.equalsIgnoreCase(call.getFnName())) {
+                return false;
+            }
+
+            if (!aggCallMapBelow.get(ref).getFnName().equalsIgnoreCase(call.getFnName())) {
+                return false;
+            }
+        }
+
+        return true;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:93]: Whether to check the distinct attributes of functions, for example, 
`select sum(ID_int) from (select sum(distinct ID_int) ID_int from test_basic) a;`

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoAggRule.java:93]: Yeah，you are right.

**Review by kangkaisen — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**